tests/integration/cmd_app_test.go

```diff
@@ -27,102 +23,137 @@ var _ = Describe("odo app command tests", func() {
-			helper.CopyExample(filepath.Join("source", "nodejs"), commonVar.Context)
-			helper.CmdShouldPass("odo", "component", "create", "--s2i", "nodejs", cmpName, "--app", appName, "--project", commonVar.Project, "--context", commonVar.Context)
-			helper.CmdShouldPass("odo", "push", "--context", commonVar.Context)
+	When("creating a new component", func() {
 
-			// list should pass as the project exists
-			appListOutput := helper.CmdShouldPass("odo", "app", "list", "--project", commonVar.Project)
-			Expect(appListOutput).To(ContainSubstring(appName))
-			helper.CmdShouldFail("odo", "app", "describe", "--project", commonVar.Project)
-			helper.CmdShouldFail("odo", "app", "delete", "-f", "--project", commonVar.Project)
-		})
-	})
+		appName := "app"
+		cmpName := "nodejs"
 
-	Context("when running app command app parameter in directory that doesn't contain .odo config directory", func() {
-		It("should successfully execute list, describe and delete along with machine readable output", func() {
+		BeforeEach(func() {
 			helper.CopyExample(filepath.Join("source", "nodejs"), commonVar.Context)
 			helper.CmdShouldPass("odo", "component", "create", "--s2i", "nodejs", cmpName, "--app", appName, "--project", commonVar.Project, "--context", commonVar.Context)
-			helper.CmdShouldPass("odo", "push", "--context", commonVar.Context)
-
-			appListOutput := helper.CmdShouldPass("odo", "app", "list", "--project", commonVar.Project)
-			Expect(appListOutput).To(ContainSubstring(appName))
-			actualCompListJSON := helper.CmdShouldPass("odo", "app", "list", "-o", "json", "--project", commonVar.Project)
-			valuesList := gjson.GetMany(actualCompListJSON, "kind", "items.#.metadata.name", "items.#.metadata.namespace")
-			expectedList := []string{"List", "app", commonVar.Project}
-			Expect(helper.GjsonMatcher(valuesList, expectedList)).To(Equal(true))
-
-			helper.CmdShouldPass("odo", "app", "describe", appName, "--project", commonVar.Project)
-			actualDesAppJSON := helper.CmdShouldPass("odo", "app", "describe", appName, "--project", commonVar.Project, "-o", "json")
-			valuesDes := gjson.GetMany(actualDesAppJSON, "kind", "metadata.name", "metadata.namespace")
-			expectedDes := []string{"Application", appName, commonVar.Project}
-			Expect(helper.GjsonMatcher(valuesDes, expectedDes)).To(Equal(true))
-
-			helper.CmdShouldPass("odo", "app", "delete", appName, "--project", commonVar.Project, "-f")
 		})
 
-	})
+		When("odo push is executed", func() {
+
+			BeforeEach(func() {
+				helper.CmdShouldPass("odo", "push", "--context", commonVar.Context)
+			})
+
+			It("should fail", func() {
+				By("describing app with app parameter", func() {
+					helper.CmdShouldFail("odo", "app", "describe", "--project", commonVar.Project)
+				})
+
+				By("deleting an app without app parameter", func() {
+					helper.CmdShouldFail("odo", "app", "delete", "-f", "--project", commonVar.Project)
+				})
+			})
```

Comment:
Hi @feloy, this part does not make much sense to me. If I was to run a test, as far as I know, I would not see the By part, so the test output would be something like -  `odo app command tests on a fresh new project [It] should fail`, which would not exactly make sense. I understand this approach would make the tests faster, but then I also feel that it contradicts with the approach you implemented in #4770 which made tests easier to read and extend.

Reply:
Good catch, you are right, I thank the By would be displayed.

I'm trying to find a compromise between the approach of #4770 and the duration of tests. Using By can reduce the execution time by running all the "BeforeEach"s only once, but it contradicts the spirit of the approach, to have only one Expectation per test.

I will make separate "It"'s instead of using "By"s, and see how longe are the tests. Running the tests on 4 nodes (or more) would however accelerate the tests